File: ProyectoCalculadora/app/src/main/java/com/example/proyectocalculadora/MiArrayAdapter.java
package com.example.proyectocalculadora;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.List;

public class MiArrayAdapter extends ArrayAdapter<Funcion> {
    public MiArrayAdapter(Context context, Funcion [] libros){
        super(context, 0, libros);
    }

    public MiArrayAdapter(Context context, List<Funcion> libros){
        super(context, 0, libros);
    }


    @Override
    public int getItemViewType(int position) {
        return (getItem(position).getExpresion() == null ? 1:0);
    }

    @Override
    public int getViewTypeCount() {
        return 2;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {

        Funcion l = getItem(position);
        View v = null;
        ViewHolder vh = null;
        TextView tvNombre = null;
        TextView tvExpresion = null;
        if(convertView!=null) {
            if((((ViewHolder)convertView.getTag()).tvExpresion == null) && (l.getExpresion() == null)){
                v = convertView;
                vh = (ViewHolder)v.getTag();
                //tvNombre = vh.tvNombre;
                vh.tvNombre.setText(l.getNombre());
            }
            else if((((ViewHolder)convertView.getTag()).tvExpresion != null) && (l.getExpresion() != null)){
                v = convertView;
                vh = (ViewHolder)v.getTag();
                /*tvExpresion = vh.tvExpresion;
                tvNombre = vh.tvNombre;*/
                vh.tvNombre.setText(l.getNombre());
                vh.tvExpresion.setText(l.getExpresion());
            }

        }
        else {
            //  IMPORTANTE : el tercer parametro esta a false porque se veria dos veces ya que el list view ya introduce el elemento por defecto.
            if(l.getExpresion() == null)
                v = LayoutInflater.from(getContext()).inflate(android.R.layout.simple_list_item_1, parent, false);
            else
                v = LayoutInflater.from(getContext()).inflate(android.R.layout.simple_list_item_2, parent, false);

            vh = new ViewHolder();
            tvNombre = (TextView) v.findViewById(android.R.id.text1);
            vh.tvNombre = tvNombre;
            tvNombre.setText(l.getNombre());
            if(l.getExpresion() != null ){
                tvExpresion = (TextView) v.findViewById(android.R.id.text2);
                vh.tvExpresion = tvExpresion;
                tvExpresion.setText(l.getExpresion());
            }
            else
                tvExpresion=null;

            v.setTag(vh);
        }//getLayoutInflater() ==> layouts



        return v;
    }

    //Clase que  sera nuestro viewHolder y que contendra las vistas mas frecuentes.
    private class ViewHolder{
        public TextView tvExpresion;
        public TextView tvNombre;

    }

}

File: ProyectoCalculadora/app/src/main/java/com/example/proyectocalculadora/MiArrayAdapter.kt
package com.example.proyectocalculadora

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.TextView

class MiArrayAdapter : ArrayAdapter<Funcion> {

    constructor(context: Context, funciones: Array<Funcion>) : super(context, 0, funciones) {}

    constructor(context: Context, funciones: List<Funcion>) : super(context, 0, funciones) {}

    override fun getItemViewType(position: Int): Int {
        return if (getItem(position)!!.expresion == null) 1 else 0
    }

    override fun getViewTypeCount(): Int {
        return 2
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View? {

        val funcion = getItem(position)
        var v: View? = null
        var vh: ViewHolder? = null
        var tvTitulo: TextView? = null
        var tvExpresion: TextView? = null
        if (convertView != null) {
            if ((convertView.tag as ViewHolder).tvExpresion == null && funcion!!.expresion == null) {
                v = convertView
                vh = v.tag as ViewHolder
                vh.tvTitulo!!.text = funcion.nombre
            } else if ((convertView.tag as ViewHolder).tvExpresion != null && funcion!!.expresion != null) {
                v = convertView
                vh = v.tag as ViewHolder
                vh.tvTitulo!!.text = funcion.nombre
                vh.tvExpresion!!.text = funcion.expresion
            }

        } else {
            //  IMPORTANTE : el tercer parametro esta a false porque se veria dos veces ya que el list view ya introduce el elemento por defecto.
            if (funcion!!.expresion == null)
                v = LayoutInflater.from(context).inflate(android.R.layout.simple_list_item_1, parent, false)
            else
                v = LayoutInflater.from(context).inflate(android.R.layout.simple_list_item_2, parent, false)

            vh = ViewHolder()
            tvTitulo = v!!.findViewById<View>(android.R.id.text1) as TextView
            vh.tvTitulo = tvTitulo
            tvTitulo.text = funcion.nombre
            if (funcion.expresion != null) {
                tvExpresion = v.findViewById<View>(android.R.id.text2) as TextView
                vh.tvExpresion = tvExpresion
                tvExpresion.text = funcion.expresion
            } else
                tvExpresion = null

            v.tag = vh
        }



        return v
    }

    //Clase que  sera nuestro viewHolder y que contendra las vistas mas frecuentes.
    private inner class ViewHolder {
        var tvTitulo: TextView? = null
        var tvExpresion: TextView? = null

    }
}
